### PR TITLE
feat(client): children for spacer component 

### LINKED ISSUE #49828: Make spacer component smarter by accepting children
We have instances which `Spacer` is used to provide spacing for both side of the component to wrap it up in specific space

Example: 

https://github.com/freeCodeCamp/freeCodeCamp/blob/49ab3866d77a0c4704a2070dcfe9e610ea5e3d6d/client/src/components/profile/profile.tsx#L125-L135

We can make the Spacer component smarter, and allow it to accept children, so the code be like.

```ts
<Spacer size='medium' > 
 {isLocked ? renderMessage(isSessionUser, username, t) : null} 
 {!isLocked || isSessionUser ? renderProfile(user) : null} 
 {isSessionUser ? null : ( 
   <Row className='text-center'> 
     <Link to={`/user/${username}/report-user`}> 
       {t('buttons.flag-user')} 
     </Link> 
   </Row> 
 )} 
 </Spacer>
```

But in doing so, we need to:

- change the current values and how the padding is calculated, the old layout shouldn't change with the new spacers.
- change which padding has the new value, I think paddingBlock, which is paddingTop and paddingBottom, is the new padding property 🤔.

---

I am requesting this because this will make live simpler, and we have to worry about less `div` elements while controling the layout, and if we have specific cases, we can still use `<Spacer />`

### PR DESCRIPTION
**Pull request by Vansh-Baghel**
… deefault

Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Ref #49828

* Added children props in the Spacer component.
* Made it optional, as we might not need it everywhere.
* Replaced the padding property with paddingBlock.
* Added additional div in case where children props exists, because if it does the paddingBlock would act as 15px top and 15px bottom of the container in case of medium. On the other hand, when children doesn't exist, it acts as 30px on the same line by calculating 15px top and 15px bottom.
* Tested it on /donate page, and it works as expected. Would implement it wherever needed in some other issue to avoid cumbersome.
* Have tested it personally on GitPod which works fine.

### CODE REVIEW

**Comment by moT01** [comment]
Thanks for working on this @Vansh-Baghel 👍 The tests need to be passing here. Mind taking a look at those?

**Comment by Vansh-Baghel** [comment]
> as far as i can see- the reason for test failure is `snapshots` that need to be updated it can be done by running client tests with `-u` parameter

Hey, I didnt understand what does *running client tests with `-u` parameter* mean, could you help me with it?

**Comment by naomi-lgbt** [comment]
@Vansh-Baghel if you run `pnpm run test-client -u` locally (while checked out to this branch), Jest will update the snapshots to reflect your changes. You'll want to commit and push those new snapshots.

**Comment by Vansh-Baghel** [comment]
> @Vansh-Baghel if you run `pnpm run test-client -u` locally (while checked out to this branch), Jest will update the snapshots to reflect your changes. You'll want to commit and push those new snapshots.

Some checks are still failing. What must I do?

File: client/src/pages/__snapshots__/email-sign-up.test.js.snap [hunk 1]
```diff
@@ -125,14 +121,13 @@ exports[`<EmailSignUp /> Non-Authenticated user "not accepted terms and conditio
...
-            "height": "1px",
-            "padding": "5px 0",
+            "paddingBlock": "5px",
           }
         }
       />
       <a
         className="big-cta-btn btn btn-lg btn-primary btn-block"
-        href="http://localhost:3000/signin"
+        href="https://3000-vanshbaghel-freecodecam-prrzb6nwdat.ws-us95.gitpod.io/signin"
```
Comment (Sboonny) [inline client/src/pages/__snapshots__/email-sign-up.test.js.snap:130]:
```suggestion
        href="http://localhost:3000/signin"
```

**Comment by Vansh-Baghel** [comment]
@Sboonny Thanks for the commit.

File: client/src/components/helpers/spacer.tsx [hunk 2]
```diff
@@ -10,13 +10,15 @@ const Padding = Object.freeze({
...
-const Spacer = ({ size }: SpacerProps): JSX.Element => (
-  <div
-    className='spacer'
-    style={{ padding: `${Padding[size]}px 0`, height: '1px' }}
-  />
+const Spacer = ({ size, children }: SpacerProps): JSX.Element => (
+  <div className='spacer' style={{ paddingBlock: `${Padding[size]}px` }}>
+    {children && (
+      <div style={{ paddingBlock: `${Padding[size]}px` }}>{children}</div>
+    )}
```
Comment (Sboonny) [inline client/src/components/helpers/spacer.tsx:20]: although this gives the desired looks, this is a `if` statement and another `div` without the need to that users have to pay for that load which isn't necessary, you can update the values in `Padding` instead.

```suggestion
    {children}
```

**Review by Sboonny — COMMENTED** [review]
I have changed the PR from closing to Ref, because there a places that has spacers and not calculating the new padding

image

File: client/src/pages/donate.tsx [hunk 3]
```diff
@@ -67,51 +67,51 @@ function DonatePage({
...
-                </Col>
-              </Row>
-            </Col>
-            <Col lg={6}>
-              <CampersImage pageName='donate' />
-            </Col>
-          </>
-        </Row>
-        <Spacer size='medium' />
+        <Spacer size='medium'>
```
Comment (Sboonny) [inline client/src/pages/donate.tsx:70]:
```suggestion
        <Spacer size='large'>
```

**Comment by Vansh-Baghel** [comment]
> I have changed the PR from closing to Ref, because there a places that has spacers and not calculating the new padding
> 
> image

I made sure to check the implementation. I checked it on donate page. In the screenshot shared by you, I couldn't find those components directly on any urls. I assumed that if one page (donate page) works, then all might work.

File: client/src/components/helpers/spacer.tsx (hunk 2 above)
Comment (Vansh-Baghel) [reply client/src/components/helpers/spacer.tsx]: That's correct!! It could be done in the previous step itself. It works!

**Comment by Sboonny** [comment]
Oh, I didn't mean you are affecting these components.

To close the issue, every `Spacer` component that can accept children should accept children, without breaking the flow of `paddingBlock`.

File: client/src/pages/donate.tsx (hunk 3 above)
Comment (Vansh-Baghel) [reply client/src/pages/donate.tsx:70]: I am not sure if it means to change it from medium to large, but doing the following change.

File: client/src/components/helpers/spacer.tsx [hunk 4]
```diff
@@ -10,13 +10,18 @@ const Padding = Object.freeze({
...
 
-const Spacer = ({ size }: SpacerProps): JSX.Element => (
+const Spacer = ({ size, children }: SpacerProps): JSX.Element => (
   <div
     className='spacer'
-    style={{ padding: `${Padding[size]}px 0`, height: '1px' }}
-  />
+    style={{
+      paddingBlock: children ? `${Padding[size] * 2}px` : `${Padding[size]}px`
+    }}
```
[the remark this answers is not in the document]
Comment (Vansh-Baghel) [reply client/src/components/helpers/spacer.tsx:21]: I see. I'll do the needful changes so it doesn't affect any future implementations.

[the remark this answers is not in the document]
Comment (Vansh-Baghel) [reply client/src/components/helpers/spacer.tsx:21]: So don't we need paddingBlock which was the point to change the calculation?

[the remark this answers is not in the document]
Comment (Sboonny) [reply client/src/components/helpers/spacer.tsx]: > So don't we need paddingBlock which was the point to change the calculation? Are we going back to padding implementation again?

no need

> We cannot put 2xl as key, as it contains an integer. Any other suggestion?

Yeab, keep it 👍

**Review by Sboonny — APPROVED** [review]
LGTM 🎉, thank you for your contribution to the page! 👍.

**Review by naomi-lgbt — DISMISSED** [review]
The snapshots still need to be updated. Running `pnpm run test-client -u` and committing+pushing the changes should allow the CI to pass.

**Comment by Vansh-Baghel** [comment]
> The snapshots still need to be updated. Running `pnpm run test-client -u` and committing+pushing the changes should allow the CI to pass.

Done fixing the snapshots!

**Review by Sembauke — APPROVED** [review]

**Pull request merged**